geoportal/c2cgeoportal_geoportal/lib/oidc.py

```diff
@@ -96,6 +97,115 @@ class OidcRememberObject(TypedDict):
+                    "Field '%s' not found in user info, available: %s.",
+                    user_info_field,
+                    ", ".join(user_info.keys()),
+                )
+                raise HTTPInternalServerError(f"Field '{user_info_field}' not found in user info.")
+            remember_object[field_] = user_info[user_info_field]  # type: ignore[literal-required]
+
+
+def get_user_from_remember(
+    request: pyramid.request.Request, remember_object: OidcRememberObject, create_user: bool = False
+) -> Union["static.User", DynamicUser] | None:
+    """
+    Create a user from the remember object filled from `get_remember_from_user_info`.
+
+    :param remember_object: The object to fill, by default with the `username`, `email`, `settings_role` and `roles`.
+    :param settings: The OpenID Connect configuration.
+    :param create_user: If the user should be created if it does not exist.
+    """
+    from c2cgeoportal_commons import models  # pylint: disable=import-outside-toplevel
+    from c2cgeoportal_commons.models import main, static  # pylint: disable=import-outside-toplevel
+
+    assert models.DBSession is not None
+
+    user: static.User | DynamicUser | None
+    username = remember_object["username"]
+    assert username is not None
+    email = remember_object["email"]
+    assert email is not None
+    display_name = remember_object["display_name"] or email
+
+    openid_connect_configuration = request.registry.settings.get("authentication", {}).get(
+        "openid_connect", {}
+    )
+    provide_roles = openid_connect_configuration.get("provide_roles", False)
+    if provide_roles is False:
+        user_query = models.DBSession.query(static.User)
+        match_field = openid_connect_configuration.get("match_field", "username")
+        if match_field == "username":
+            user_query = user_query.filter_by(username=username)
+        elif match_field == "email":
+            user_query = user_query.filter_by(email=email)
+        else:
+            raise HTTPInternalServerError(
+                f"Unknown match_field: '{match_field}', allowed values are 'username' and 'email'"
+            )
+        user = user_query.one_or_none()
+        if user is not None:
+            for field in openid_connect_configuration.get("update_fields", []):
+                if field == "username":
+                    user.username = username
+                elif field == "display_name":
+                    user.display_name = display_name
+                elif field == "email":
+                    user.email = email
+                else:
+                    raise HTTPInternalServerError(
+                        f"Unknown update_field: '{field}', allowed values are 'username', 'display_name' and 'email'"
+                    )
+        if user is None and create_user is True:
+            user = static.User(username=username, email=email, display_name=display_name)
+            models.DBSession.add(user)
```

Comment:
Shouldn't this section be in the login function? This will create a transaction for each request otherwise.